vowpalwabbit/lda_core.cc

```diff
@@ -809,12 +858,227 @@ void learn(lda &l, LEARNER::base_learner &, example &ec)
     learn_batch(l);
 }
 
+void learn_with_metrics(lda &l, LEARNER::base_learner &base, example &ec)
+{
+  if (l.all->passes_complete == 0)
+  { // build feature to example map
+    auto weight_mask = l.all->reg.weight_mask;
+    auto stride_shift = l.all->reg.stride_shift;
+    
+    for (features& fs : ec)
+    { for (features::iterator& f : fs)
+      { uint64_t idx = (f.index() & weight_mask) >> stride_shift;
+        l.feature_counts[idx] += f.value();
+        l.feature_to_example_map[idx].push_back(ec.example_counter);
+      }
+    }
+  }
+
+  learn(l, base, ec);
+}
+
 // placeholder
 void predict(lda &l, LEARNER::base_learner &base, example &ec) { learn(l, base, ec); }
+void predict_with_metrics(lda &l, LEARNER::base_learner &base, example &ec) { learn_with_metrics(l, base, ec); }
+
+struct word_doc_frequency
+{
+	uint64_t idx;
+	uint32_t count;
+};
+
+struct feature_pair
+{
+	uint64_t f1;
+	uint64_t f2;
+
+	feature_pair(uint64_t _f1, uint64_t _f2) : f1(_f1), f2(_f2)
+	{}
+};
+
+void compute_metrics(lda &l)
```

Comment:
Perhaps add comments / clean up this part to make it more readable. Additionally, should this be named something like compute_coherence?